src/rust/engine/src/externs/address.rs

```diff
@@ -1,21 +1,786 @@
 // Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+use std::borrow::Cow;
+use std::collections::{BTreeMap, HashSet};
+use std::ffi::OsStr;
+use std::hash::{Hash, Hasher};
+use std::path::{Path, PathBuf};
+
+use pyo3::basic::CompareOp;
 use pyo3::create_exception;
-use pyo3::exceptions::PyException;
+use pyo3::exceptions::{PyAssertionError, PyException};
 use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyFrozenSet, PyType};
+
+use fnv::FnvHasher;
+use lazy_static::lazy_static;
 
 create_exception!(native_engine, AddressParseException, PyException);
+create_exception!(native_engine, InvalidAddressError, AddressParseException);
+create_exception!(native_engine, InvalidSpecPathError, InvalidAddressError);
+create_exception!(native_engine, InvalidTargetNameError, InvalidAddressError);
+create_exception!(native_engine, InvalidParametersError, InvalidAddressError);
+create_exception!(native_engine, UnsupportedWildcardError, InvalidAddressError);
 
 pub fn register(py: Python, m: &PyModule) -> PyResult<()> {
+  m.add_function(wrap_pyfunction!(address_spec_parse, m)?)?;
+
   m.add(
     "AddressParseException",
     py.get_type::<AddressParseException>(),
   )?;
-  m.add_function(wrap_pyfunction!(address_spec_parse, m)?)?;
+  m.add("InvalidAddressError", py.get_type::<InvalidAddressError>())?;
+  m.add(
+    "InvalidSpecPathError",
+    py.get_type::<InvalidSpecPathError>(),
+  )?;
+  m.add(
+    "InvalidTargetNameError",
+    py.get_type::<InvalidTargetNameError>(),
+  )?;
+  m.add(
+    "InvalidParametersError",
+    py.get_type::<InvalidParametersError>(),
+  )?;
+  m.add(
+    "UnsupportedWildcardError",
+    py.get_type::<UnsupportedWildcardError>(),
+  )?;
+
+  m.add_class::<AddressInput>()?;
+  m.add_class::<Address>()?;
+
+  m.add(
+    "BANNED_CHARS_IN_TARGET_NAME",
+    PyFrozenSet::new(py, BANNED_CHARS_IN_TARGET_NAME.iter())?,
+  )?;
+  m.add(
+    "BANNED_CHARS_IN_GENERATED_NAME",
+    PyFrozenSet::new(py, BANNED_CHARS_IN_GENERATED_NAME.iter())?,
+  )?;
+  m.add(
+    "BANNED_CHARS_IN_PARAMETERS",
+    PyFrozenSet::new(py, BANNED_CHARS_IN_PARAMETERS.iter())?,
+  )?;
+
   Ok(())
 }
 
+lazy_static! {
+  // `:`, `#`, `@` are used as delimiters already. Others are reserved for possible future needs.
+  pub static ref BANNED_CHARS_IN_TARGET_NAME: HashSet<char> =
+    [':', '#', '!', '@', '?', '/', '\\', '='].into();
+  pub static ref BANNED_CHARS_IN_GENERATED_NAME: HashSet<char> =
+    [':', '#', '!', '@', '?', '='].into();
+  pub static ref BANNED_CHARS_IN_PARAMETERS: HashSet<char> =
+    [':', '#', '!', '@', '?', '=', ',', ' '].into();
+}
```

Comment:
These are so small that I wonder whether linear search (e.g. `[':', '#', ...]` arrays or `":#..."` strings) would be faster, and allow saving avoiding the hashset. Doesn't actually matter, though.